substrate/client/network/sync/src/chain_sync.rs

```diff
@@ -1212,7 +1200,11 @@ where
 
 		let blocks = self.ready_blocks();
 
-		(!blocks.is_empty()).then(|| self.validate_and_queue_blocks(blocks, false))
+		if !blocks.is_empty() {
+			let ImportBlocksAction { origin, blocks } =
+				self.validate_and_queue_blocks(blocks, false);
+			self.actions.push(ChainSyncAction::ImportBlocks { origin, blocks })
```

Comment:
Why is this not happening inside of `validate_and_queue_blocks`?

Reply:
Fixed — now all actions are pushed at the place they are generated. Except `BadPeer` errors — those are sometime deferred until the function returns to simplify the error handling.

Reply:
Well, actually there is one function left returning actions — `new_peer_inner`.  But we need to track in `restart` that it is not requesting any actions to cancel obsolete requests, so it's hard to make it not return actions.

Reply:
Still need to fix tests :see_no_evil:

Reply:
Now unit tests are not that unit anymore...

Reply:
The tests are updated, but they became uglier because of the need to track actions pushed into `self.actions` instead of just checking return values.
@bkchr could you have a look again?

Reply:
I'm going to merge this to move forward, but we can revisit these changes later if you know better ways of testing structs with an output queue.